### ISSUE TITLE
libopen-pal.so: undefined reference to PMIx_Value_unload/PMIx_Info_xfer/PMIx_Info_load

### ISSUE THREAD

**Issue by kenorb** [open-mpi/ompi#10047]
### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)

```
$ git rev-parse HEAD
aeafbde1900c78f746888bf2ca0d75a491d10e18
```

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

```
./configure --disable-dlopen
make
```

### If you are building/installing from a git clone, please copy-n-paste the output from `git submodule status`.

```
 b98b7e83519963f7ba484ea34449668ea6834a0e 3rd-party/openpmix (v1.1.3-3376-gb98b7e83)
 1717de3cf3d1c079d1f55480ca5a14f20d0e723c 3rd-party/prrte (psrvr-v2.0.0rc1-4203-g1717de3cf3)
```

### Please describe the system on which you are running

* Operating system/version: Ubuntu 21.10
* Computer hardware: PC
* Network type: WiFi

-----------------------------

## Details of the problem

During `make` (configured with `--disable-dlopen`), I've got the following errors:

```
  CCLD     opal_wrapper
/usr/bin/ld: ../../../opal/.libs/libopen-pal.so: undefined reference to `PMIx_Value_unload'
/usr/bin/ld: ../../../opal/.libs/libopen-pal.so: undefined reference to `PMIx_Info_xfer'
/usr/bin/ld: ../../../opal/.libs/libopen-pal.so: undefined reference to `PMIx_Info_load'
make[2]: *** [Makefile:1476: opal_wrapper] Error 1
```

**Comment by kenorb** [comment]
I think it works for v4.1.2 tag.

**Comment by kenorb** [comment]
I think I've found the problem, the `./configure` failed with the following errors (which I must to miss it before):

> configure: error: Could not find PMIx devel headers.  Can not continue.

```
$ git checkout master
$ git rev-parse HEAD
0f50e03d656a86883fb6a742e73f3d0a94c9b8b2
$ ./autogen.pl
...
Open MPI autogen: completed successfully.  w00t!
...
$ ./configure 
...
== Configure PMIx
...
checking src/util/pmix_argv.h usability... no
checking src/util/pmix_argv.h presence... no
checking for src/util/pmix_argv.h... no
configure: error: Could not find PMIx devel headers.  Can not continue.
configure: ===== done with 3rd-party/prrte configure =====
configure: error: PRRTE configuration failed.  Cannot continue.
```

I've `libpmix-dev` installed fine.

```
$ sudo apt show libpmix-dev
Package: libpmix-dev
Version: 4.1.0-3ubuntu1
```

Al thought the mentioned files are in submodule at https://github.com/openpmix/openpmix/blob/94169f1725a26a156fd73fca1b61df14ccd581ab/src/util/pmix_argv.h

So it seems configure is not checking them correctly.

I'm attaching my `config.log`: [config.log.gz](https://github.com/open-mpi/ompi/files/8191215/config.log.gz)

Lines of interest:

```
conftest.c:(.text.startup+0xb): undefined reference to `PMIx_Init'
collect2: error: ld returned 1 exit status
```

**Comment by jsquyres** [comment]
There's unfortunately not enough information in the `config.log` you sent -- Open MPI's `configure` script invokes the sub-`configure` scripts in the embedded copies of `PMIx` and `PRTE` (which then generate their own `config.log` files).  Can you send the full stdout/stderr from running `configure`?

Note that it's expected that some of `configure`'s tests will fail.  It's therefore not unusual to see failures in `config.log`.  That being said, PRTE failing to configure is obviously a fatal error.

However, I can't quite reconcile your original post with your later comment:

1. The original post was about how building Open MPI failed because it couldn't find the PMIx symbols when it tried to link an Open MPI executable.
2. [A later comment](https://github.com/open-mpi/ompi/issues/10047#issuecomment-1059810223) specified how `configure` failed.

If `configure` failed, no Makefiles would have been generated, and you would not have been able to invoke `make`.  Hence, I think you're talking about two different errors here.

As a sidenote, is there a reason you're building Open MPI from git, rather than using a distribution tarball from https://www.open-mpi.org/software/ompi/v4.1/?  (v4.1.x is the current stable release series)

Note, too, that per #10074, there was a bad PMIx submodule update this past Friday on Open MPI's `master` branch that may be causing at least some build failures.  This likely own't get fixed until tomorrow (Monday) at the earliest.

**Comment by kenorb** [comment]
> However, I can't quite reconcile your original post with your later comment

Most likely different errors happened at different recent HEADs.

- `aeafbde1900c78f746888bf2ca0d75a491d10e18` - ./configure passed (didn't verify the missing headers), so build failed during the compilation (missing PMIx references)
- `0f50e03d656a86883fb6a742e73f3d0a94c9b8b2` (latest master) - ./configure fails before finding missing PMIx devel headers

I'm going to wait for the further fixes and re-test.

I did test today with the latest master `0f50e0` (from scratch) and it fails at the `./configure`.

**Comment by jsquyres** [comment]
Looks like #10074 has been fixed (got some new commits from the submodule).  The problem would have occurred on machines with the netlink library headers.

**Comment by kenorb** [comment]
I've tested the latest `eb9b9c42e179f13ba8b86d06e895e557f3b89fe2`, and I've the same error:

```
$ ./configure
...
checking for src/util/pmix_argv.h... no
configure: error: Could not find PMIx devel headers.  Can not continue.
configure: ===== done with 3rd-party/prrte configure =====
configure: error: PRRTE configuration failed.  Cannot continue.

$ git rev-parse HEAD
eb9b9c42e179f13ba8b86d06e895e557f3b89fe2
$ git submodule status
 94169f1725a26a156fd73fca1b61df14ccd581ab 3rd-party/openpmix (v1.1.3-3429-g94169f17)
 6722bd767e0fccbe07193f59f4f13e5610c7dd6a 3rd-party/prrte (psrvr-v2.0.0rc1-4274-g6722bd767e)
$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working tree clean
```

**Comment by jsquyres** [comment]
Is there a reason you're not using an official release tarball?

The output I can see on the screenshot of the video says why it failed: it found a PMIx install in /usr/local, but it couldn't find the PMIx devel headers.

Did you previously install PMIx into /usr/local without the devel headers?  If so, you have two choices:

1. Uninstall the PMIx in /usr/local and just use the PMIx obtained via git submodule in the Open MPI git clone (all the Right Magic will happen to use the embedded PMIx if configure can't find an external PMIx).
2. Re-install PMIx to /usr/local with the devel headers.  You'll have to check how to do that; I think it's configuring `--with-devel-headers`, or somesuch.

If your goal is just to write some MPI apps, option 1 is definitely simpler.  There's no need to install PMIx separately (ditto for PRTE, but that's only relevant for Open MPI's master branch / what will become Open MPI v5.0).

**Comment by kenorb** [comment]
> Is there a reason you're not using an official release tarball?

Yes, I could use v4.1.2 instead which configures and compiles fine.

Regarding `master` branch, I've tried to provide the necessary information so you could solve the issue in the latest code, but it could be specific to my local environment. Maybe it found previous PMIx headers which I've installed from v4.1.2 branch.

- I did remove `libpmix-dev` and `libpmix2` Apt packages, and reconfigured, didn't change the errors for master.
- I've removed `pmix` and `openmpi` from `/usr/local` as suggested (previously installed from another branch), by:

      sudo rm -frv /usr/local/**/openmpi/** /usr/local/**/pmix/**

- I've run `sudo make uninstall` from `v4.1.2` tag, but it seems this version didn't have submodules at all.

Then re-cloned repo from scratch (`master`), and reconfigured, the same errors are still there.

I could continue with v4.1.2 version. If you need further information, please let me know.

**Comment by jsquyres** [comment]
Thanks for taking the time to submit everything.  This is _probably_ an issue with your local environment.

From your last output, I see you've removed a bunch of PMIx stuff such that pmix.pc (the PMIx pkg-config file) is no longer found, but you didn't include the rest of the configure output that shows why it still thinks it found PMIx but no devel headers.

**Comment by bwbarrett** [comment]
It would be very helpful to have the entire output of OMPI's configure.  It's hard to figure out exactly what is happening from the short snippets.
